Refine value ranges on inequalities

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #114148
* __->__ #120800

This is basically done the obvious way. For better or worse, I jammed this into what used to be `_maybe_guard_eq` but now is `_maybe_guard_rel`. I was careful to test all the off by one conditions, and each permutation. Let me know if you think I missed anything. Importantly, this now works for unbacked SymInts.

While testing, I noticed we are silently duck sizing all symbolic variables in `test_dynamic_shapes.py`. This may or may not be covering up bugs.

Along the way, I had to fix a bug in export constraints, where we weren't checking that the final var_to_range was consistent with what the user requested at top level.

After I implemented all this, I realized that applying this to non-unbacked SymInts was duplicative with @ysiraichi's previous work on https://github.com/pytorch/pytorch/pull/97963 . The upside is I now understand what Yukio was trying to do in the original PR, and I think my new logic is simpler and less error prone. In Yukio's earlier diff, Yukio tried very hard to avoid changing what guards we actually issue (since this would cause tests to wobble). Thus, when he refined a range, he also saved the guard that actually caused the range to refine. In this PR, I don't bother saving these guards; instead I just tighten var_to_range directly and rely on generating guards on this to be correct. The key insight is that if I assert `x < y`, it's always safe to emit (potentially) more restrictive range guards, because this won't invalidate our guards, it will just make them a little too strong (but actually, I think we are precise along the way.) If these guards make it unnecessary to test `x < y`, because now the ranges for x and y are disjoint, this is fine, we've subsumed the x < y guard and can just not bother testing it. If I've gotten it right, TV will agree with me.

In fact, I had a bug in this PR which TV didn't catch, which is that when we have a recorded var_to_guards for a symbol, we unconditionally never generate the range guard for it, even if the var_to_guards is potentially inconsistent with var_to_range (because var_to_range was updated separately). With var_to_guards removed, I don't have to worry abou this inconsistency.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames

ezyang committed:
Update

[ghstack-poisoned]

ezyang committed:
Update

[ghstack-poisoned]

Skylion007 commented on test/test_dynamic_shapes.py:
```diff
@@ -554,6 +555,70 @@ def test_expect_true_double_digits(self):
 
+    def test_expect_true_refine_range(self):
+        shape_env = ShapeEnv()
+        for rel in [lambda x: x > 4, lambda x: 4 < x, lambda x: x >= 5, lambda x: 5 <= x]:
```
Nit: probably want to add some 'with self.subTest' in this for loop to make debugging easier.

ezyang replied:
bah, then I need to name the lambdas, someone else can add it if they need it LOL

ezyang committed:
Update

[ghstack-poisoned]

ezyang committed:
Update

[ghstack-poisoned]

ezyang committed:
Update

[ghstack-poisoned]

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3069,40 +3073,62 @@ def issue_guard(guard: ShapeGuard) -> None:
+                                c.warn_only,
+                                self._debug_name(source),
+                                msg,
+                            )
```
@avikchaudhuri new constraint checking code.

avikchaudhuri replied:
lgtm! 

I'd like to rationalize `sympy.oo` vs. `sys.maxint - 1` (for export I think we encode upper bounds as the latter but even if we didn't, they become the latter when they go into `var_to_range`) so I don't think you'll see `sympy.oo` here when doing `c.vr & ...` for export errors; but good to have it anyway. Also FYI we haven't done `0` lower bound yet, @pianpwk will probably do it soon. 

Curious: what's the status of `specialize_zero_one`, does size-oblivious have any bearing on it? Asking because `_default_value_range` switches based on it.

ezyang replied:
I actually had to add the `&` to fix an export test lol, so it's definitely needed (although I feel like the current `>= 2` constraint was the most laod bearing.

Re `specialize_zero_one`, this is kind of... very misleading. What this does is say if the online solver should *eagerly* specialize on 0/1 quantities (and make the negative inference when the hint is not these). In practice, you will pretty quickly end up with this specialization anyway because of Tensor constructor guards. But maybe sometimes you wouldn't, and toggling this off would help you be less specialized in that case.

ezyang committed:
Update on "Refine value ranges on inequalities"


This is basically done the obvious way. For better or worse, I jammed this into what used to be `_maybe_guard_eq` but now is `_maybe_guard_rel`. I was careful to test all the off by one conditions, and each permutation. Let me know if you think I missed anything.

While testing, I noticed we are silently duck sizing all symbolic variables in `test_dynamic_shapes.py`. This may or may not be covering up bugs.

Along the way, I had to fix a bug in export constraints, where we weren't checking that the final var_to_range was consistent with what the user requested at top level.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

avikchaudhuri commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3069,40 +3073,62 @@ def issue_guard(guard: ShapeGuard) -> None:
+            assert symbol.is_integer
+            g_lower, g_upper = self.var_to_guards.get(symbol, (None, None))
+            bounds = []
+            if r.lower != -sympy.oo and g_lower is None:
```
`r.lower` can be `sympy.oo`?

ezyang replied:
Yes, when you have an unbacked symint that is not known to be size-like

avikchaudhuri commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3665,6 +3674,43 @@ def _maybe_guard_eq(self, expr: Union["sympy.Eq", "sympy.Ne"], concrete_bool: bo
         lhs = expr.lhs
         rhs = expr.rhs
+
+        if not is_eq:
```
Is the equality case handled elsewhere? Do we just merge value ranges with `&` for them? Or is that unsafe?

ezyang replied:
The equality stuff is the pre-existing code in this function, below. It indirectly updates value ranges because when we are able to solve the equality into a replacement, the variable is eliminated and so you get updated value ranges.

Your comment here makes me realize that in principle, we could do better on multi-variate expressions by trying to progressively solve for each variable and propagate ranges for each one we are able to solve, but IDK, seems cursed.

lezcano replied:
That would propagate ranges nicely I think. One should do it until it finds a fixed point, but that should only take 1-2 iters at most.

ezyang committed:
Update on "Refine value ranges on inequalities"


This is basically done the obvious way. For better or worse, I jammed this into what used to be `_maybe_guard_eq` but now is `_maybe_guard_rel`. I was careful to test all the off by one conditions, and each permutation. Let me know if you think I missed anything. Importantly, this now works for unbacked SymInts.

While testing, I noticed we are silently duck sizing all symbolic variables in `test_dynamic_shapes.py`. This may or may not be covering up bugs.

Along the way, I had to fix a bug in export constraints, where we weren't checking that the final var_to_range was consistent with what the user requested at top level.

After I implemented all this, I realized that applying this to non-unbacked SymInts was duplicative with ysiraichi's previous work on https://github.com/pytorch/pytorch/pull/97963 . The upside is I now understand what Yukio was trying to do in the original PR, and I think my new logic is simpler and less error prone. In Yukio's earlier diff, Yukio tried very hard to avoid changing what guards we actually issue (since this would cause tests to wobble). Thus, when he refined a range, he also saved the guard that actually caused the range to refine. In this PR, I don't bother saving these guards; instead I just tighten var_to_range directly and rely on generating guards on this to be correct. The key insight is that if I assert `x < y`, it's always safe to emit (potentially) more restrictive range guards, because this won't invalidate our guards, it will just make them a little too strong (but actually, I think we are precise along the way.) If these guards make it unnecessary to test `x < y`, because now the ranges for x and y are disjoint, this is fine, we've subsumed the x < y guard and can just not bother testing it. If I've gotten it right, TV will agree with me.

In fact, I had a bug in this PR which TV didn't catch, which is that when we have a recorded var_to_guards for a symbol, we unconditionally never generate the range guard for it, even if the var_to_guards is potentially inconsistent with var_to_range (because var_to_range was updated separately). With var_to_guards removed, I don't have to worry abou this inconsistency.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

ezyang committed:
Update on "Refine value ranges on inequalities"


This is basically done the obvious way. For better or worse, I jammed this into what used to be `_maybe_guard_eq` but now is `_maybe_guard_rel`. I was careful to test all the off by one conditions, and each permutation. Let me know if you think I missed anything. Importantly, this now works for unbacked SymInts.

While testing, I noticed we are silently duck sizing all symbolic variables in `test_dynamic_shapes.py`. This may or may not be covering up bugs.

Along the way, I had to fix a bug in export constraints, where we weren't checking that the final var_to_range was consistent with what the user requested at top level.

After I implemented all this, I realized that applying this to non-unbacked SymInts was duplicative with ysiraichi's previous work on https://github.com/pytorch/pytorch/pull/97963 . The upside is I now understand what Yukio was trying to do in the original PR, and I think my new logic is simpler and less error prone. In Yukio's earlier diff, Yukio tried very hard to avoid changing what guards we actually issue (since this would cause tests to wobble). Thus, when he refined a range, he also saved the guard that actually caused the range to refine. In this PR, I don't bother saving these guards; instead I just tighten var_to_range directly and rely on generating guards on this to be correct. The key insight is that if I assert `x < y`, it's always safe to emit (potentially) more restrictive range guards, because this won't invalidate our guards, it will just make them a little too strong (but actually, I think we are precise along the way.) If these guards make it unnecessary to test `x < y`, because now the ranges for x and y are disjoint, this is fine, we've subsumed the x < y guard and can just not bother testing it. If I've gotten it right, TV will agree with me.

In fact, I had a bug in this PR which TV didn't catch, which is that when we have a recorded var_to_guards for a symbol, we unconditionally never generate the range guard for it, even if the var_to_guards is potentially inconsistent with var_to_range (because var_to_range was updated separately). With var_to_guards removed, I don't have to worry abou this inconsistency.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

lezcano commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3639,22 +3639,31 @@ def _find(self, a: "sympy.Symbol") -> "sympy.Expr":
-    def _maybe_guard_eq(self, expr: Union["sympy.Eq", "sympy.Ne"], concrete_bool: bool) -> None:
+    def _maybe_guard_rel(self, expr: "sympy.Rel", concrete_bool: bool) -> None:
         """
         Evaluates the result of an eq call. If true, uses information to
```
eq -> relational

lezcano reviewed (commented):
The `_maybe_guard_rel` and the tests lgtm

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -4115,18 +4158,15 @@ def _refine_ranges(self, guard: ShapeGuard) -> None:
 
             # Updates the range and the guards corresponding to each bound of the symbol.
             self.var_to_range[symbol] = ValueRanges(lower, upper)
-            self.var_to_guards[symbol] = (lower_guard, upper_guard)
```
The reason it is not necessary (and you can see all the tests passed!) is because we issue dedicated guards for the value range. So in your example, when you have range `[6, ...]` we produce a guard `s0 >= 6`. If I discard a subsequent guard, including `s0 > 5`, it is because they were made strictly redundant by the range specific guard.

When I originally reviewed your PR, what I didn't quite understand was how to deal with things like `s0 < s1` which showed up in `var_to_guards`. Could these be necessary to preserve? But in fact, I have convinced myself that it is not necessary. If this guard is not statically implied by the refined ranges, I won't drop it (as you are worried). If it is statically implied, it is not needed, the other guards + var ranges are enough.

ysiraichi replied:
That makes a lot of sense. Thank you for explaining it to me thoroughly.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3665,6 +3674,43 @@ def _maybe_guard_eq(self, expr: Union["sympy.Eq", "sympy.Ne"], concrete_bool: bo
+            # we can't put s0 in the range.
+            #
+            # TODO: deal with conjunction/disjunction?
+            if len(free) == 1:
+                lhs = free[0]
+                r = try_solve(expr, lhs)
+                if r is not None and isinstance(r[0], (sympy.Lt, sympy.Le, sympy.Gt, sympy.Ge)) and isinstance(r[1], sympy.Integer):
+                    res = r[0]
+                    bound = int(r[1])
+                    if isinstance(res, sympy.Lt):
+                        # lhs < bound
+                        update_vr = [-sympy.oo, bound - 1]
+                    elif isinstance(res, sympy.Le):
+                        # lhs <= bound
+                        update_vr = [-sympy.oo, bound]
+                    elif isinstance(res, sympy.Gt):
+                        # lhs > bound
+                        update_vr = [bound + 1, sympy.oo]
+                    elif isinstance(res, sympy.Ge):
+                        # lhs >= bound
+                        update_vr = [bound, sympy.oo]
+                    else:
+                        raise AssertionError(f"illegal expr {expr}")
+                    old_range = self.var_to_range[lhs]
+                    self.var_to_range[lhs] &= ValueRanges(*update_vr)
+                    if old_range != self.var_to_range[lhs]:
+                        self.log.info("refined range on %s to %s", lhs, self.var_to_range[lhs])
+                        self._maybe_evaluate_static.cache_clear()
```
It is basically the same, and a reasonable BE followup for me would be to eliminate one or the other. Honestly, the only reason there's a second copy is I didn't realize this was here. There are two major differences with the implementations:

* `refine_ranges` attempts to solve for every free variable, whereas `maybe_guard_rel` only propagates in the univariate case. This could potentially be a performance problem, especially when you have guards like `s0 + ... + s999 = s1000`, but it also increases precision. In the unbacked case, I actually do have a lot of guards like this, so I think the performance concern here is warranted.
* `refine_ranges` handles non-integral RHS (e.g., `x < f(y)`) by computing the bound of `f(y)` and then using the conservative end of the range. This is unconditionally good and worth using.

ezyang committed:
Update on "Refine value ranges on inequalities"


This is basically done the obvious way. For better or worse, I jammed this into what used to be `_maybe_guard_eq` but now is `_maybe_guard_rel`. I was careful to test all the off by one conditions, and each permutation. Let me know if you think I missed anything. Importantly, this now works for unbacked SymInts.

While testing, I noticed we are silently duck sizing all symbolic variables in `test_dynamic_shapes.py`. This may or may not be covering up bugs.

Along the way, I had to fix a bug in export constraints, where we weren't checking that the final var_to_range was consistent with what the user requested at top level.

After I implemented all this, I realized that applying this to non-unbacked SymInts was duplicative with ysiraichi's previous work on https://github.com/pytorch/pytorch/pull/97963 . The upside is I now understand what Yukio was trying to do in the original PR, and I think my new logic is simpler and less error prone. In Yukio's earlier diff, Yukio tried very hard to avoid changing what guards we actually issue (since this would cause tests to wobble). Thus, when he refined a range, he also saved the guard that actually caused the range to refine. In this PR, I don't bother saving these guards; instead I just tighten var_to_range directly and rely on generating guards on this to be correct. The key insight is that if I assert `x < y`, it's always safe to emit (potentially) more restrictive range guards, because this won't invalidate our guards, it will just make them a little too strong (but actually, I think we are precise along the way.) If these guards make it unnecessary to test `x < y`, because now the ranges for x and y are disjoint, this is fine, we've subsumed the x < y guard and can just not bother testing it. If I've gotten it right, TV will agree with me.

In fact, I had a bug in this PR which TV didn't catch, which is that when we have a recorded var_to_guards for a symbol, we unconditionally never generate the range guard for it, even if the var_to_guards is potentially inconsistent with var_to_range (because var_to_range was updated separately). With var_to_guards removed, I don't have to worry abou this inconsistency.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

ezyang commented:
Updated to use Yukio's `refine_ranges` directly instead of duplicating the logic

ezyang committed:
Update on "Refine value ranges on inequalities"


This is basically done the obvious way. For better or worse, I jammed this into what used to be `_maybe_guard_eq` but now is `_maybe_guard_rel`. I was careful to test all the off by one conditions, and each permutation. Let me know if you think I missed anything. Importantly, this now works for unbacked SymInts.

While testing, I noticed we are silently duck sizing all symbolic variables in `test_dynamic_shapes.py`. This may or may not be covering up bugs.

Along the way, I had to fix a bug in export constraints, where we weren't checking that the final var_to_range was consistent with what the user requested at top level.

After I implemented all this, I realized that applying this to non-unbacked SymInts was duplicative with ysiraichi's previous work on https://github.com/pytorch/pytorch/pull/97963 . The upside is I now understand what Yukio was trying to do in the original PR, and I think my new logic is simpler and less error prone. In Yukio's earlier diff, Yukio tried very hard to avoid changing what guards we actually issue (since this would cause tests to wobble). Thus, when he refined a range, he also saved the guard that actually caused the range to refine. In this PR, I don't bother saving these guards; instead I just tighten var_to_range directly and rely on generating guards on this to be correct. The key insight is that if I assert `x < y`, it's always safe to emit (potentially) more restrictive range guards, because this won't invalidate our guards, it will just make them a little too strong (but actually, I think we are precise along the way.) If these guards make it unnecessary to test `x < y`, because now the ranges for x and y are disjoint, this is fine, we've subsumed the x < y guard and can just not bother testing it. If I've gotten it right, TV will agree with me.

In fact, I had a bug in this PR which TV didn't catch, which is that when we have a recorded var_to_guards for a symbol, we unconditionally never generate the range guard for it, even if the var_to_guards is potentially inconsistent with var_to_range (because var_to_range was updated separately). With var_to_guards removed, I don't have to worry abou this inconsistency.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge -i
